### ISSUE TITLE
Stanford CS224N 2020 - Low Resource Machine Translation

### ISSUE THREAD

**Issue by kimziwoo** [kimziwoo/paper-summaries#1]
_lectured by Marc' Aurelio Ranzato, Facebook AI Research - NYC_
_all images copyright stanford CS224N_
_reordered and renumbered by me_ 
_자세한 기법에 대한 분석은 페이퍼 리뷰와 함께할 예정입니다_

# 0. 강의 도입부 : Introduce NMT
### Low-Resource in Language

전세계에는 6000개가 넘는 언어가 있고, 전세계 인구의 80%가 영어를 사용하지 않으며 5%의 미만 사람들만 영어를 모국어로 사용한다. 2000개가 넘는 언어가 1000명이 안되는 사람들에게 발화된다. 

<img src="https://user-images.githubusercontent.com/86106509/131292795-823c0b8d-ea14-428a-bd75-a0130e4d09eb.png" width="60%" height="50%"></img>

### Low-Resource in Domain 

low-resource는 언어뿐만 아니라 도메인에도 해당된다.  위 그림(몬드리안 learning setting)처럼, 다른 언어의 다른 도메인으로 구성된 데이터셋이 typical하다. OPUS만 보더라도 종종 out-domain으로 데이터셋이 구성된다는 점을 알 수 있다. 

## 0.2 LR-MT's Challenges
- loose def: when the number of parallel sentences is in the order of 10,000 or less. 

LR-MT의 challenge는 다음과 같다. 

```
1. data 측면 
- sourcing data to train on, evaluation datasets. 
특히 lecturer는 evaluation, test set이 부족한 점을 강조해서 얘기했는데, 이후 FLoRes Evaluation Benchmark(EMNLP 
2019)에 대해서 소개했다. FLoRes는 English-Nepali와 English-Sinhala에서 각각 3000개의 문장으로 이뤄진 validation, test and hidden test set임.

2. modeling 측면
- unclear learning paradigm, domain adaptation, generalization
```

## 0.3 Before ML Perspective - Case-study : En-Ne background
lecturer는 English-Nepali 상황을 가정해서 여러 ML perspective들을 소개했다.   

먼저 data에 대한 이해를 위해 아래 그림 하단 설명을 보자. 이 지점에서 전체 domain별로 살피다가 잠시 헷갈렸는데, **test data가 어딨는지를 기준**으로 삼으면 설명이 된다. wikipedia 도메인에 있는 EN-NE test data는 parallel data가 없고 monolingual data만 갖고 있다. **parallel data와 monolingual data가 대치된다**는 점에 주목하자. 또한 bible, GNOME 도메인의 Nepali 데이터는 empty 사각형으로 되어있는데, 이는 English 데이터를 human-translate한 걸 의미한다. 이때는 parallel한 데이터이므로 해당 도메인 내에서(bible, GNOME in-domain) parallel data를 갖고있지만, 기준은 test data이므로 out-of-domain에서의 parallel data를 갖고있다.   

한편 human translation data는 원재료가 해당 나라에서 나오지 않았으므로, Nepali에서부터 originate한 번역은 없다고 할 수 있다(test data 기준). 

<img src="https://user-images.githubusercontent.com/86106509/131294599-3c01b44f-48e5-4060-8288-5e52bfa8e232.png" width="60%" height="50%"></img>

ML perspective에 대해 자세한 건 다음 절에서부터 설명. 

# 1. ML Perspective 소개: Supervised Learning

주어진 데이터셋은 다음과 같다. 

![image](https://user-images.githubusercontent.com/86106509/131303860-55daa8eb-f08b-474e-b441-5fcc1f3883ba.png)

아주 일반적인 상황이다. 

![image](https://user-images.githubusercontent.com/86106509/131306408-98b77456-c0a5-42a9-8a5b-0655bc756295.png)
![image](https://user-images.githubusercontent.com/86106509/131306489-cbd4542f-bd27-4de2-8884-43f8fd631e01.png)

따라서 일반적으로 attention-based transformer로 학습시킨다. 

![image](https://user-images.githubusercontent.com/86106509/131306339-7f1d87cf-87c0-4803-b0e6-cbf6c4b36246.png)

하지만 low-resource data는 training dataset의 N이 너무 작은 문제가 발생하는데, 이때는 주로 다음과 같은 방법을 사용한다. 
- dropout
- label smoothing 

# 2. ML Perspective 소개: Semi-Supervised Learning

![image](https://user-images.githubusercontent.com/86106509/131307211-6c48b793-d129-48e1-8326-d58609b0ecb6.png)

![image](https://user-images.githubusercontent.com/86106509/131307981-2f8f31b8-e4da-4aa9-9892-95ee6a02e59d.png)

위와 같은 데이터셋이 주어졌다. **즉 source-side에 monolingual data가 추가적으로 주어진 상황**. 어떻게 이 데이터(M_s)를 활용할 수 있을까? 

## 2-1. DAE(Denoising Auto-Encoders)

![image](https://user-images.githubusercontent.com/86106509/131307871-52d645c9-4557-4556-b5ee-8e9ff923d055.png)

활용하는 방법 첫번째, DAE loss를 pretrain 하거나 기존 supervised cross-entropy term에 추가하면 된다. 

![image](https://user-images.githubusercontent.com/86106509/131307897-b2a83ab8-e975-4f43-a2b2-275a4a407e97.png)

여기서 noise는, 다음과 같은 방법으로 추가한다.
- word drop
- swap
- etc ...

## 2-2. ST(self-training)

![image](https://user-images.githubusercontent.com/86106509/131308777-ceba6f5a-d9b4-4175-845c-8ee627b68917.png)

활용하는 방법 두번째, Self-Training 방법이다. 
다음과 같은 루틴으로 이뤄진다. 

```
1. train on machine translation system on the parallel data. 
2. repeat the following process. 
2-1. decode monolingual dataset using current machine translation system. 
2-2. make a new parallel data set of sentences from monolingual dataset with translation from current system
2-3. retrain the model. 

```

정리한 알고리즘은 다음과 같다. 

![image](https://user-images.githubusercontent.com/86106509/131308905-a3be6fa3-c7f9-47f0-a8d9-5270321301dd.png)

## 2-3 BT(Back-Translation)

![image](https://user-images.githubusercontent.com/86106509/131316221-cbd90dbb-8379-4e2e-a011-53ec31551dd5.png)

위와 같이 만약 source-side가 아닌 target-side에 monolingual data가 있다면 Back-Translation 기법을 사용하면 된다. 

![image](https://user-images.githubusercontent.com/86106509/131316271-ee3dbd95-6087-471e-8320-b48419493e34.png)

알고리즘은 다음과 같다. 

![image](https://user-images.githubusercontent.com/86106509/131316541-780c6971-d9da-40a7-9a1f-a6b1a9b0da98.png)

이는 다음과 같은 이점을 가진다. 

```

1. Decoder learns a good language model. 
2. Better generalization via data augmentation. 
3. Unlike ST, target is correct but input is not. 

```

## 2-4 Iterative ST & BT

![image](https://user-images.githubusercontent.com/86106509/131316881-c8a795e3-db9e-4296-9b0d-bceb2509c0d3.png)

![image](https://user-images.githubusercontent.com/86106509/131316956-49aaaf3b-107b-4d69-9f61-685db8a826f2.png)

위와 같이 source-side와 target-side 모두 monolingual한 데이터만 있다면(parallel data 없는 경우) 다음과 같은 방법을 사용할 수 있다. 

![image](https://user-images.githubusercontent.com/86106509/131316981-d2ba1b92-783d-4e2d-8975-cb070ea5a0e4.png)

알고리즘은 다음과 같다. 

![image](https://user-images.githubusercontent.com/86106509/131317014-4760c1b2-47d5-48a6-967d-edcd695f0c85.png)

# 3. ML Perspective: Multi-Task/Multi-Modal

![image](https://user-images.githubusercontent.com/86106509/131318105-8efd9413-0682-4937-b7b9-d168dd95a0ff.png)

![image](https://user-images.githubusercontent.com/86106509/131318137-78181556-747a-45e8-a390-9fb7777b750d.png)

이외에도 multilingual learning framework가 있다.   
모든 language pair가 같은 encoder와 decoder를 공유하는 구조로, 모든 데이터셋을 함께 concat한다.   
일반적인 cross-entropy loss를 사용한다. 

## In practical

lecturer가 말하기를, 현시점엔(2020) 앞서 나온 ST, BT, Multilingual training, DAE 등의 기법을 얼마나 잘 combine하는지에 따라 성능이 좌우된다고 한다.   

![image](https://user-images.githubusercontent.com/86106509/131319304-401a5284-7330-480b-9292-09b34d06cb73.png)

예시로 lecturer는 Unsupervised MT로 monolingual dataset만 주어진 경우를 제시했는데, 이는 Iterative BT와 DAE와 Multi-Lingual의 조합으로 해결할 수 있다. 

![image](https://user-images.githubusercontent.com/86106509/131319464-f215ebb8-9e5c-4f6c-8de6-2e689bd4be53.png)

Iterative BT

![image](https://user-images.githubusercontent.com/86106509/131319517-e854d2b0-f8a8-4592-bb4c-a15f12ac8291.png)

DAE

![image](https://user-images.githubusercontent.com/86106509/131319542-e269f896-8ec4-47ee-a448-fa1460cf302c.png)

Multi-Lingual 

# conclusion so far 

lecturer는 다음과 같은 결론을 내리며 마무리한다. 

- Iterative back-translation과 multi-lingual training work remarkably well. 
- **By feeding more data(BT, ST, pre-training, multi-lingual training) we can afford training bigger models. Bigger models train on more data generalize better.** 
- Low-resource MT requires big compute! Remember that about 100 monolingual sentences give the same training signal as a single pair of parallel sentence. 

기억해야할 점은 "low-resource"라고 해서 작은 데이터와 작은 모델이 아니라는 것이다. 적은 데이터를 보충(overfitting 방지)하기 위해서 더 많은 데이터가 필요하고, 그를 위해 source-side와 target-side의 monolingual 데이터를 적극 활용해야 한다. 그에 따라 complicated learning setting을 이해하고 조립해야한다. 

# later? 
- Back-translation
- Self-training
- regularization 관련 공부
- multilingual learning
- DAE 등 Auto Encoder 관련 공부
